src/dials/algorithms/correlation/analysis.py

```diff
@@ -0,0 +1,233 @@
+from dials.util.filter_reflections import filtered_arrays_from_experiments_reflections
+from dials.util.multi_dataset_handling import select_datasets_on_identifiers
+from dials.util.observer import Subject
+
+logger = logging.getLogger("dials.algorithms.correlation.analysis")
+
+phil_scope = iotbx.phil.parse(
+    """\
+partiality_threshold = 0.4
+  .type = float
+  .help = "Use reflections with a partiality above the threshold."
+
+include scope dials.algorithms.symmetry.cosym.phil_scope
+
+relative_length_tolerance = 0.05
+  .type = float(value_min=0)
+
+absolute_angle_tolerance = 2
+  .type = float(value_min=0)
+
+min_reflections = 10
+  .type = int(value_min=1)
+  .help = "The minimum number of reflections per experiment."
+""",
+    process_includes=True,
+)
+
+
+class CorrelationMatrix(Subject):
+    def __init__(self, experiments, reflections, params=None):
+        """
+        Set up the required cosym preparations for determining the correlation matricies
+        of a series of input experiments and reflections.
+        Args:
+          experiments (dxtbx_model_ext.ExperimentList): dials experiments
+          reflections (list): dials reflections
+          params (libtbx.phil.scope_extract): experimental parameters
+        """
+
+        if params is None:
+            params = phil_scope.extract()
+        self.params = params
+
+        self._reflections = []
+        for refl, expt in zip(reflections, experiments):
+            sel = get_selection_for_valid_image_ranges(refl, expt)
+            self._reflections.append(refl.select(sel))
+
+        self._experiments, self._reflections = self._filter_min_reflections(
+            experiments, self._reflections
+        )
+        self.ids_to_identifiers_map = {}
+        for table in self._reflections:
+            self.ids_to_identifiers_map.update(table.experiment_identifiers())
+        self.identifiers_to_ids_map = {
+            value: key for key, value in self.ids_to_identifiers_map.items()
+        }
+
+        datasets = filtered_arrays_from_experiments_reflections(
+            self._experiments,
+            self._reflections,
+            outlier_rejection_after_filter=False,
+            partiality_threshold=params.partiality_threshold,
+        )
+        individual_merged_intensities = []
+        for unmerged in datasets:
+            individual_merged_intensities.append(
+                unmerged.merge_equivalents().array().set_info(unmerged.info())
+            )
+        self.datasets = [
+            d.eliminate_sys_absent(integral_only=True).primitive_setting()
+            for d in individual_merged_intensities
+        ]
+
+        self.params.lattice_group = self.datasets[0].space_group_info()
+        self.params.space_group = self.datasets[0].space_group_info()
+
+        self.cosym_analysis = CosymAnalysis(self.datasets, self.params)
+
+    def _filter_min_reflections(self, experiments, reflections):
```

Comment:
This is broken out as a separate method but feels like that could be done to some advantage with other bits as well?